Add logback tracing to Java LS

snjeza opened the pull request:
Fixes #2108 

You can test in the following way
- add -Djdt.ls.debug=true to java.jdt.ls.vmargs
- open a maven project
- check the Java LS server log

You can override it using -Dlogback.configurationFile=<your-logback.xml>

Notice: Now, m2e.core requires Java-17
- https://github.com/eclipse-m2e/m2e-core/pull/740
- https://github.com/eclipse-m2e/m2e-core/commit/01a638721b315179de996daf2d40fd6d1ffb8262
Java LS tests fail because of that.

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

snjeza commented:
> Is it possible to combine jdt.ls.logback into jdt.ls.start to have just 1 extra bundle ?

No, it isn't. jdt.ls.start have to start first so can't be a fragment. 

> Could the [JavaLanguageServerStart.java](https://github.com/eclipse/eclipse.jdt.ls/pull/2116/files#diff-2261cbe91fb67c8c923791f56a843757014ed9a806033ba9ed0b6de6d921ec4e) not be added inside jdt.ls.core since we no longer need to call setProperty very early ?

No, it couldn't. We have to call setProperty very early if exists ch.qos.logback

testforstephen commented:
Having a separate start plugin to set logback property looks like kind of hack. I don't get the point why we cannot do it in jdt.ls bundle.

snjeza commented:
> Having a separate start plugin to set logback property looks like kind of hack. I don't get the point why we cannot do it in jdt.ls bundle.

@testforstephen You may want to take a look at https://github.com/eclipse/eclipse.jdt.ls/pull/2096#issuecomment-1141489175

rgrunber commented:
Also, for the nicer solution of logging directly to Eclipse's logs, I think the logback framework only checks its classpath for custom appenders, so effectively that means we need to use `Fragment-Host`. Not sure we can do that for jdt.ls.core since a fragment is not considered a bundle with its own Activator start/stop. See https://github.com/eclipse-m2e/m2e-core/blob/master/org.eclipse.m2e.logback.appender/META-INF/MANIFEST.MF#L8 , and the custom appender at https://github.com/eclipse-m2e/m2e-core/blob/master/org.eclipse.m2e.logback.appender/src/org/eclipse/m2e/logback/appender/EclipseLogAppender.java . Unfortunately we can't use that because of the UI dependencies.

snjeza commented:
test this please

rgrunber commented:
Just one observation about the logging. Importing a project like eclipse/lemminx I see about 15 messages going into the server logs. However, I remember seeing many more when I was able to enable verbose/debug logging. For example :

[.out-jdt.ls-20220610140605.log](https://github.com/eclipse/eclipse.jdt.ls/files/8881170/default.out-jdt.ls-20220610140605.log)

It's a bit excessive, but is it possible to have control over the DEBUG messages also ? i think they're the ones that are currently missing. I though adding a `case Level.DEBUG_INT:` in `EclipseLogAppender` would solve it but it doesn't seem to have worked.

snjeza commented:
> It's a bit excessive, but is it possible to have control over the DEBUG messages also ? i think they're the ones that are currently missing. I thought adding a case Level.DEBUG_INT: in EclipseLogAppender would solve it but it doesn't seem to have worked.

checking...

snjeza commented:
@rgrunber Could you, please, check again?

snjeza commented:
test this please

snjeza commented:
test this please

snjeza commented:
test this please

snjeza commented:
test this please

snjeza commented:
test this please

snjeza commented:
test this please

snjeza commented:
test this please

rgrunber commented:
> > Seems to work consistently.
> 
> Could you try to create a vsix with Java 17 jre and start a maven project in a new workspace? See [#2096 (comment)](https://github.com/eclipse/eclipse.jdt.ls/pull/2096#issuecomment-1141418008)
> 
> > Note that, I don't mind regressing on having the ability to supply a custom logback file through system properties. It was a nice-to-have feature, but ultimately, the most important thing is just being able to log the data with some sane defaults
> 
> You can try to set -Dlogback.configurationFile=<your_logback.xml> jvmargs. It should work.

You're right. It would break that case.

testforstephen commented on org.eclipse.jdt.ls.start/src/org/eclipse/jdt/ls/start/JavaLanguageServerStart.java:
```diff
@@ -0,0 +1,144 @@
+	private static final String LOGBACK_CONFIG_FILE_PROPERTY = "logback.configurationFile";
+	private static final String LOGBACK_DEFAULT_FILENAME = "logback.xml";
+	private static final String LOGBACK_EMPTY_FILENAME = "emptylogback.xml";
+	public static final String PLUGIN_ID = "org.eclipse.jdt.ls.start";
```
If we have to create a new bundle to solve this problem, I would rename it from "org.eclispe.jdt.ls.start" to "**org.eclipse.jdt.ls.logback**", and make this bundle only used for handling logback logic.

snjeza replied:
Fixed.

testforstephen commented on org.eclipse.jdt.ls.start/src/org/eclipse/jdt/ls/start/JavaLanguageServerStart.java:
```diff
@@ -0,0 +1,144 @@
+			// ch.qos.logback.classic.util.ContextInitializer.CONFIG_FILE_PROPERTY
+			// System.setProperty(LOGBACK_CONFIG_FILE_PROPERTY, configFile.getAbsolutePath());
+		}
+		try {
+			Platform.getBundle(CORE_PLUGIN_ID).start(Bundle.START_TRANSIENT);
+		} catch (BundleException e) {
+			logException(e.getMessage(), e);
+		}
```
Why do we need to start jdt.ls.core bundle manually here? It looks unnecessary for me.

snjeza replied:
You are right. I have removed it.

snjeza commented:
@rgrunber @testforstephen I have updated the PR. I have also updated the target platform to 4.24 because 4.24-I-builds doesn't exist.
re failing tests, see #2123

snjeza commented:
test this please

testforstephen commented:
> @testforstephen , so currently, the m2e logs are only appended to the `.log` (JDT-LS runtime logs) file when `-Djdt.ls.debug=true` is added to `java.jdt.ls.vmargs`. So this change shouldn't add any more data to the logs by default, as one must explicitly add the setting.
> 
> on loading a project like spring-petclinic : Regular logging (no m2e logging) prints about 16KB (197 lines) for me Debug logging (with m2e logging) prints about 220KB (3509 lines) for me

Good to know it. Since it won't append to the .log file by default, so it's not a problem for language server daemon.

testforstephen commented:
When I execute "Java: Clean Java Language Server Workspace" commad and reload window, it reports the following exception.

```
!ENTRY org.eclipse.jdt.ls.logback 4 0 2022-06-17 11:25:55.056
!MESSAGE C:\Users\myName\AppData\Roaming\Code\User\workspaceStorage\c2d6260b73da6e426b91d7bd9c5b1036\redhat.java\jdt_ws\.metadata\.plugins\org.eclipse.jdt.ls.logback\emptylogback.xml: The process cannot access the file because it is being used by another process
!STACK 0
java.nio.file.FileSystemException: C:\Users\myName\AppData\Roaming\Code\User\workspaceStorage\c2d6260b73da6e426b91d7bd9c5b1036\redhat.java\jdt_ws\.metadata\.plugins\org.eclipse.jdt.ls.logback\emptylogback.xml: The process cannot access the file because it is being used by another process
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:275)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:110)
	at java.base/java.nio.file.Files.deleteIfExists(Files.java:1191)
	at java.base/java.nio.file.Files.copy(Files.java:3148)
	at org.eclipse.jdt.ls.logback.JavaLanguageServerStart.start(JavaLanguageServerStart.java:66)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:818)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:810)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:767)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1032)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:371)
	at org.eclipse.osgi.container.Module.doStart(Module.java:605)
	at org.eclipse.osgi.container.Module.start(Module.java:468)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1847)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor$1$1.execute(EquinoxContainerAdaptor.java:136)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1840)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1783)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1745)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1667)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:345)

!ENTRY org.eclipse.jdt.ls.logback 4 0 2022-06-17 11:25:55.152
!MESSAGE FrameworkEvent ERROR
!STACK 0
org.osgi.framework.BundleException: Exception in org.eclipse.jdt.ls.logback.JavaLanguageServerStart.start() of bundle org.eclipse.jdt.ls.logback.
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:839)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:767)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1032)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:371)
	at org.eclipse.osgi.container.Module.doStart(Module.java:605)
	at org.eclipse.osgi.container.Module.start(Module.java:468)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1847)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor$1$1.execute(EquinoxContainerAdaptor.java:136)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1840)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1783)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1745)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1667)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:345)
Caused by: ch.qos.logback.core.joran.spi.JoranException: Problem parsing XML document. See previously reported errors.
	at ch.qos.logback.core.joran.event.SaxEventRecorder.recordEvents(SaxEventRecorder.java:65)
	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:151)
	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:110)
	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:53)
	at org.eclipse.jdt.ls.logback.JavaLanguageServerStart.loadConfiguration(JavaLanguageServerStart.java:110)
	at org.eclipse.jdt.ls.logback.JavaLanguageServerStart.start(JavaLanguageServerStart.java:71)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:818)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:810)
	... 14 more
Caused by: org.xml.sax.SAXParseException; systemId: file:/C:/Users/jinbwan.FAREAST/AppData/Roaming/Code/User/workspaceStorage/c2d6260b73da6e426b91d7bd9c5b1036/redhat.java/jdt_ws/.metadata/.plugins/org.eclipse.jdt.ls.logback/emptylogback.xml; lineNumber: 1; columnNumber: 1; Premature end of file.
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1251)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:326)
	at ch.qos.logback.core.joran.event.SaxEventRecorder.recordEvents(SaxEventRecorder.java:59)
	... 23 more
Root exception:
ch.qos.logback.core.joran.spi.JoranException: Problem parsing XML document. See previously reported errors.
	at ch.qos.logback.core.joran.event.SaxEventRecorder.recordEvents(SaxEventRecorder.java:65)
	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:151)
	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:110)
	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:53)
	at org.eclipse.jdt.ls.logback.JavaLanguageServerStart.loadConfiguration(JavaLanguageServerStart.java:110)
	at org.eclipse.jdt.ls.logback.JavaLanguageServerStart.start(JavaLanguageServerStart.java:71)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:818)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:810)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:767)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1032)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:371)
	at org.eclipse.osgi.container.Module.doStart(Module.java:605)
	at org.eclipse.osgi.container.Module.start(Module.java:468)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1847)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor$1$1.execute(EquinoxContainerAdaptor.java:136)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1840)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1783)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1745)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1667)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:345)
Caused by: org.xml.sax.SAXParseException; systemId: file:/C:/Users/jinbwan.FAREAST/AppData/Roaming/Code/User/workspaceStorage/c2d6260b73da6e426b91d7bd9c5b1036/redhat.java/jdt_ws/.metadata/.plugins/org.eclipse.jdt.ls.logback/emptylogback.xml; lineNumber: 1; columnNumber: 1; Premature end of file.
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1251)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:326)
	at ch.qos.logback.core.joran.event.SaxEventRecorder.recordEvents(SaxEventRecorder.java:59)
	... 23 more

!ENTRY org.eclipse.osgi 4 0 2022-06-17 11:25:55.158
!MESSAGE Bundle org.eclipse.jdt.ls.logback_1.13.0.202206170301 [54] is not active.

```

snjeza commented:
@testforstephen I can reproduce the issue on Windows. I have fixed it. Could you, please, try again.

snjeza commented:
test this please

testforstephen commented:
With the latest changes, I didn't see the exception above. But I found another issue. When I removed `-Djdt.ls.debug=true` from java.jdt.ls.vmargs, it still printed verbose m2e log to the .log file.

snjeza commented:
> With the latest changes, I didn't see the exception above. But I found another issue. When I removed -Djdt.ls.debug=true from java.jdt.ls.vmargs, it still printed verbose m2e log to the .log file.

@testforstephen I can't reproduce the issue. Could you attach your server log?

testforstephen commented:
> > With the latest changes, I didn't see the exception above. But I found another issue. When I removed -Djdt.ls.debug=true from java.jdt.ls.vmargs, it still printed verbose m2e log to the .log file.
> 
> @testforstephen I can't reproduce the issue. Could you attach your server log?

It turns out I have enabled jdt.ls.debug somewhere. After I cleaned it, no m2e log is printed to .log file.

snjeza commented:
@rgrunber @testforstephen I have updated the PR.
It uses Java ServiceProvider/ServiceLoader https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html to configure the logback preferences. 
The org.eclipse.jdt.ls.logback plugin is not needed.

testforstephen commented on org.eclipse.jdt.ls.logback.appender/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,10 @@
+Bundle-Version: 1.13.0.qualifier
+Fragment-Host: ch.qos.logback.classic
+Automatic-Module-Name: org.eclipse.jdt.ls.logback.appender
+Bundle-RequiredExecutionEnvironment: JavaSE-11
```
it should be `JavaSE-17`

testforstephen reviewed (approved):
The latest change works well for me.
